src/panels/lovelace/cards/hui-area-card.ts

```diff
@@ -0,0 +1,418 @@
+  extends SubscribeMixin(LitElement)
+  implements LovelaceCard
+{
+  public static async getConfigElement(): Promise<LovelaceCardEditor> {
+    await import("../editor/config-elements/hui-area-card-editor");
+    return document.createElement("hui-area-card-editor");
+  }
+
+  public static getStubConfig(): AreaCardConfig {
+    return { type: "area", area: "" };
+  }
+
+  @property({ attribute: false }) public hass!: HomeAssistant;
+
+  @state() private _config?: AreaCardConfig;
+
+  @state() private _entities?: EntityRegistryEntry[];
+
+  @state() private _devicesInArea: Set<string> = new Set();
+
+  // null means area not found. undefined = loading.
+  @state() private _area?: AreaRegistryEntry | null;
+
+  private _memberships = memoizeOne(
+    (
+      areaId: string,
+      devicesInArea: Set<string>,
+      registryEntities: EntityRegistryEntry[],
+      states: HomeAssistant["states"]
+    ) => {
+      const entitiesInArea = registryEntities
+        .filter(
+          (entry) =>
+            !entry.entity_category &&
+            (entry.area_id
+              ? entry.area_id === areaId
+              : entry.device_id && devicesInArea.has(entry.device_id))
+        )
+        .map((entry) => entry.entity_id);
+
+      const sensorEntities: HassEntity[] = [];
+      const entitiesToggle: HassEntity[] = [];
+
+      for (const entity of entitiesInArea) {
+        const domain = computeDomain(entity);
+        if (!TOGGLE_DOMAINS.has(domain) && !SENSOR_DOMAINS.includes(domain)) {
+          continue;
+        }
+
+        const stateObj: HassEntity | undefined = states[entity];
+
+        if (!stateObj) {
+          continue;
+        }
+
+        if (entitiesToggle.length < 3 && TOGGLE_DOMAINS.has(domain)) {
+          entitiesToggle.push(stateObj);
+          continue;
+        }
+
+        if (
+          sensorEntities.length < 3 &&
+          SENSOR_DOMAINS.includes(domain) &&
+          stateObj.attributes.device_class &&
+          SENSOR_DEVICE_CLASSES.includes(stateObj.attributes.device_class)
+        ) {
+          sensorEntities.push(stateObj);
+        }
+
+        if (sensorEntities.length === 3 && entitiesToggle.length === 3) {
+          break;
+        }
+      }
+
+      return { sensorEntities, entitiesToggle };
+    }
+  );
+
+  public hassSubscribe(): UnsubscribeFunc[] {
+    return [
+      subscribeAreaRegistry(this.hass!.connection, (areas) => {
+        this._area =
+          areas.find((area) => area.area_id === this._config!.area) || null;
+      }),
+      subscribeDeviceRegistry(this.hass!.connection, (devices) => {
+        this._devicesInArea = new Set(
+          this._config
+            ? devices
+                .filter((device) => device.area_id === this._config!.area_id)
+                .map((device) => device.id)
+            : []
+        );
+      }),
+      subscribeEntityRegistry(this.hass!.connection, (entries) => {
+        this._entities = entries;
+      }),
+    ];
```

Comment:
Doing the filtering in here means that when the config changes, (when using the editor for example) you don't have the correct devices and area.